### PR TITLE
fix #89 Add individual paste buffers to cursors

### LINKED ISSUE #89: copy&paste not working
In IntelliJ I can have multiple selections, copy them, e.g. move the cursors 3 lines down and then paste the different lines. Vim currently just pastes the last line that was selected X times.

### PR DESCRIPTION
**Pull request by balta2ar**
With this commit you can use multiple cursors to yank/paste and they will use
their own block of text as expected. In Visual mode operations `y` and `Y` are
supported. In Normal mode, all operations that change unname register are
supported (`dcsx`) but only few motions are supported (e.g. `cw`, `ciw`, `caw`,
`ci(`, and so on, but no support for `ctx` or `c/pattern<cr>`).

For best experience, use these options in your ~/.vimrc:

    let g:multi_cursor_normal_maps = {'d': 1, 'c': 1}
    let g:multi_cursor_exit_from_insert_mode = 0
    let g:multi_cursor_exit_from_visual_mode = 0

### CODE REVIEW

**Comment by balta2ar** [comment]
This hopefully fixes #89. These are my first baby steps in VimL, so somebody experienced should definitely check that. @eapache Could you please try/review this?

**Comment by balta2ar** [comment]
More tests could be added for commands with motions, such as `ciw`, 'da(', I just need to figure out how to override global options in unit tests.

**Comment by eapache** [comment]
I'm a bit busy at the moment but I will try to find time to review this sometime in the next week, this is definitely a feature I want!

**Comment by faceleg** [comment]
Me too!

File: README.md [hunk 1]
```diff
@@ -109,6 +109,20 @@ as `dw` to delete a word) work in multi-cursor mode. You have to
...
+saving/restoring paste buffer of the cursors. With this option you can select
+text in Visual mode, yank it, move the cursors, and paste it back. Each cursor
+will paste the text it yanked. `y` and `Y` are supported in Visual mode.
+Normal mode is supported but limited: `C`, `s`, `S`, `x`, `X`, `D` (commands
+without motions) work as usual. Only few of commands with motions are
+supported: `ci`, `ca`, `cb`, `ce`, `cw` and similar. `t`, `f`, `/` motions are
+not supported. Note that it is assumend that commands with motions such that
+`c` and `d` are added to `g:multi_cursor_normal_maps`.
+
+You should not change this option under normal circumstances. Setting value to
```
Comment (eapache) [inline README.md]: is there any real reason to make it configurable then?

File: autoload/multiple_cursors.vim [hunk 2]
```diff
@@ -457,9 +481,27 @@ function! s:CursorManager.update_current() dict
...
+
     call cur.remove_visual_selection()
   elseif s:from_mode ==# 'i' && s:to_mode ==# 'n' && self.current_index != self.size() - 1
     normal! h
+  elseif s:from_mode ==# 'n'
+    " Save contents of unnamed register after each supported operation in Normal
+    " mode. Paste buffer should be saved only if we are about to perform a
+    " supported operation that is known to change unnamed register. List of
+    " supported operations is fixed and defined above.
+    " This is ugly and can probably be done more elegantly and reliably.
```
Comment (eapache) [inline autoload/multiple_cursors.vim]: Hmm, what about we just checked after every operation on every cursor if the register contents have changed, and if they have, save?

File: autoload/multiple_cursors.vim [hunk 3]
```diff
@@ -54,6 +54,13 @@ if !hlexists(s:hi_group_visual)
   exec "highlight link ".s:hi_group_visual." Visual"
 endif
 
+" List of operations (later compared with s:char[0]) that should trigger saving
+" of paste buffer for each cursor. The idea is to change cursor's paste buffer
+" only when operation is in fact changing unnamed cursor. Checking like this is
+" ugly and there should be a better, reliable and more portable way to do that.
+" This list is only used for checks in Normal mode.
+let s:paste_buffer_supported_operations = 'cCsSxXdD'
```
Comment (eapache) [inline autoload/multiple_cursors.vim]: what about `y` in normal mode? or is that handle separately?

**Comment by eapache** [comment]
So I think this approach will generally work, and I'm not sure what other approaches you tried, but if I were to try writing this I think I would approach it slightly more simply:
- after each cursor operation, save the contents of the register
- before each cursor operation, restore the contents of the register

This would do unnecessary work on operations that don't copy/paste any data, but I think it would be simpler and easier to make work with e.g. the more complex motions. Thoughts?

File: README.md (hunk 1 above)
Comment (balta2ar) [inline README.md]: Well, not anymore. The history behind this option is that I didn't immediately realize I could turn `clipboard` off while cursors are active. I personally have `clipboard+=unnamed,unnamedplus` and at first I was testing without that option and unnamed `@"` register worked. But when you turn that option on, looks like `@"` is not the cursor you want anymore, and then I started using `@+` instead. That's where all sorts of weird sh-t started to happen. I've seen behaviour I couldn't explain, literally: when the last statement in a function is `let ...`, the state gets screwed up during the call, but when I add `echom ...`, it works just fine. Anyway, after I discovered that trick to save/restore clipboard that option became redundant, I'll remove it.

File: autoload/multiple_cursors.vim (hunk 3 above)
Comment (balta2ar) [inline autoload/multiple_cursors.vim]: Hm, no, it's not. I probably just forgot about that. I've just tried to add `yY` here and `y` to `normal_maps` and it seems to work. Good catch, thanks!

**Comment by balta2ar** [comment]
That was exactly my first approach :) Unfortunately I had to give it up. Here is why. You see, if we pay no attention to what cursor operation we're saving register after, operations such as `b`, `w` or any other movements in Normal mode would overwrite paste buffer of all cursors: remember that vim has only one unnamed register, thus recording its contents to all paste buffers during operation that does not change unnamed register will copy that same block of text N times which is not what we want. We *have* to save register *only* during replay of `csdxy` operations because that's *the only* moment of time when we know the value of `@"` for each cursor, after that it's lost (except for the last cursor).

But I didn't give up that easily. Next, I though, ok, let's improve that approch: why not check if current value of unnamed register is exactly equal to the value of any of paste buffers? If unnamed register is already present somewhere, then we must be just moving around, right? Wrong. Imagine the following case: first we put 3 cursors on the second column of words one bellow another and yank them:

```
send send
send receive
send skip
```

If we start moving after that, it's good, imagine last cursor yanked `send` (`@"` = `send`), which is present in at least one paste buffer. But now we put our cursors on the first column of `send` words and yank them. It doesn't work because we have one `send` already stored in one of our paste buffers.

Long story short, my method is ugly and I know it. I don't even know if it works with remappings: what if we `nnoremap h c`, will `c` be detected as a "register changing operation"? What we need is some kind of notification from vim about type of operation that is about to happen. Sadly, I'm not aware of such API, are you? Thoughts? Am I missing something?

**Comment by eapache** [comment]
Hmmm... you're right it is much trickier than I thought at first glance!

> why not check if current value of unnamed register is exactly equal to the value of any of paste buffers?

My gut says there's still an approach here that works. What about if, for each operation:
- save the current register to a temporary variable
- set the current register to the contents on the buffer (if any) 
- run the operation
- save the register to the buffer
- restore the register from the temporary variable only if it didn't change during the operation

I *think* that covers all the cases nicely?

File: README.md (hunk 1 above)
Comment (balta2ar) [inline README.md]: I completely forgot about another use of this option: if you set it to empty string, the functionality of individual paste buffers is disabled. Do you think we need that?

Comment (eapache) [inline README.md]: Unless it has bugs severe enough people would want to disable it, I don' think so :)

I can't think of any case where the current behaviour is really useful, can you?

Comment (balta2ar) [inline README.md]: What if somebody's working with gigantic blocks of text with multiples cursors and it makes memory usage very high. But that's rather thought experiment than a real case scenario, I think. OK, I'll get rid of it.

**Comment by balta2ar** [comment]
I've been desperately trying to grasp the conditon: `only if it didn't change during the operation` and failed. Instead I decided to just go and implement it without that condition. It doesn't work because register gets copied into all cursor buffers during movement operations.

Just so we are clear, let me put that into pseudocode (R = register, T = temporary, B = paste buffer):

1. T=R
2. if (B) R=B
3. operation
4. B=R
5. R=T (what's the pseudocode for `only if it didn't change during the operation`?)

**Comment by eapache** [comment]
Hmm, yes, that condition was supposed to deal with the problem you noted, but I did a very poor job of explaining. As I edited my original post to add though, I think that condition is entirely unnecessary if you make the second step unconditional as well. Working pseudocode should go something like this:

```
T=R
R=B
operation
B=R
R=T
```

Note that `R=B` should run even if `B` is empty. Also note that this means we would 'lose' the existing contents of the register when entering multiple-cursor mode; I think that can be remedied by initializing new cursor buffers to the contents of the register, rather than just an empty string.

**Comment by balta2ar** [comment]
I'm afraid it still doesn't work. As soon as I disable `if strridx(s:paste_buffer_supported_operations, s:char[0]) >= 0` check, it breaks the tests:

```
Multiple Cursors op pending & exit from insert|visual mode
  #paste buffer normal caw then p (FAILED - 1)
  #paste buffer normal C then ABC then p
  #paste buffer normal daw then P (FAILED - 2)
  #paste buffer normal D then P (FAILED - 3)
  #paste buffer normal s then p
```

I did the following changes according to your description: http://sprunge.us/WjDA?diff
Am I missing something?

**Comment by balta2ar** [comment]
If fact, I am. I forgot about this check: `if s:from_mode ==# 'n' && s:char ==? 'p'`. Now tests pass! Hooray and glory to @eapache :)

I've left a couple of comments in the last commit. Please review it all once again to check whether I didn't forget any leftovers or misleading comments or anything like that. And thank you for the simple idea, you rock! :)

**Comment by balta2ar** [comment]
> I was thinking originally that both save and restore could happen regardless of mode... It's a bit odd to me that right now you save in visual mode but only restore in normal mode...

Hmmm... I should probably run some performance/latency tests and see if it makes any difference with and without mode checks.

**Comment by eapache** [comment]
> Hmmm... I should probably run some performance/latency tests and see if it makes any difference with and without mode checks.

I've noticed a few times some latency timers built-in to the code, but I don't know how to use them.

My only *real* concern here is that operations in visual mode do multiple saves in a row without a restore, meaning that `T` gets overwritten several times and the original contents of the buffer gets lost.

One potential performance win, now I think about it, is that the original value of `R` (saved and restored from `T` all the time) should basically never change at all when in multi-cursor mode. You might not have to save/restore it on each operation, but only when entering/leaving multi-cursor mode (you would still have to save/restore the per-cursor buffer on each operation).

**Comment by eapache** [comment]
Apart from the concern about multiple-possible-saves-without-a-restore in visual mode, this looks really good - the other issues are just "nice-to-have", and since this ticket is getting quite long I suggest you start a new ticket to work on them (if you want) once this one has been submitted.

**Comment by balta2ar** [comment]
> My only real concern here is that operations in visual mode do multiple saves in a row without a restore, meaning that T gets overwritten several times and the original contents of the buffer gets lost.

I moved saving/restoring of `T` to init/exit functions, please check if I chose the right place. If I understand it right, the problem should be gone. But even before that I tried to reproduce it and I couldn't. I created N cursors, and tried to move in Visual mode, change selection. The value of unnamed register before and after multiple cursors mode was the same. Isn't it `restore` which is called in Visual mode? `save` is only called in Normal mode, see `process_user_input`.

There are other glitches that I've noticed:
* if you have something in `@"` and you start N cursors, then pastes, I expect all N cursors to paste the contents of `@"`. It pastes something else now.
* `undo` does something to the cursors that reverses some of them. If you yank, paste, see the output, then undo and paste again, output will be different.

I agree that work on those should be done in a separate ticket. If you feel like the work here is done, let's wrap up and merge this.

File: autoload/multiple_cursors.vim [hunk 4]
```diff
@@ -317,6 +325,17 @@ function! s:Cursor.remove_visual_selection() dict
   let self.visual_hi_id = 0
 endfunction
 
+" Restore unnamed register from paste buffer
+function! s:Cursor.save_paste_buffer() dict
```
Comment (eapache) [inline autoload/multiple_cursors.vim]: I think I would rename these to match the comments (so this one becomes `restore_paste_register` and the other becomes `save_paste_register`) but you might disagree

**Comment by eapache** [comment]
The `undo` problem will probably be complicated and certainly can wait. The initial paste problem is unexpected, so I'd kind of like to track that down before we merge this (we should have a test for it as well).

File: autoload/multiple_cursors.vim (hunk 4 above)
Comment (balta2ar) [inline autoload/multiple_cursors.vim]: Technically we don't have "paste register", we have "paste buffer" (our concept) and "unnamed register" (vim's concept). How about we name it `restore_unnamed_register`?

Comment (eapache) [inline autoload/multiple_cursors.vim]: :+1:

**Comment by balta2ar** [comment]
I've added test to check that unnamed register is pasted into 3 cursors. I didn't have to fix anything, looks like I accidentially fixed it in  839aed9.

**Comment by eapache** [comment]
Cool, this looks awesome, thanks for all your work!

**Comment by balta2ar** [comment]
It's been a pleasure working with you :)
